src/resources/spi_linux.cc

```diff
@@ -26,75 +27,242 @@
 #include <sys/ioctl.h>
 #include <linux/spi/spidev.h>
 
+#include "../event_sources/spi_linux.h"
+
 namespace toit {
 
+enum {
+  kTransferDone = 1 << 0,
+};
+
 // Defined in primitive_file_non_win.cc.
 extern Object* return_open_error(Process* process, int err);
 
+class SpiResourceGroup : public ResourceGroup {
+ public:
+  TAG(SpiResourceGroup);
+  explicit SpiResourceGroup(Process* process)
+    : ResourceGroup(process, SpiEventSource::instance()) {}
+
+ protected:
+  void on_unregister_resource(Resource* resource) override;
+
+  uint32 on_event(Resource* resource, word data, uint32_t state) override {
+    USE(resource);
+    state |= data;
+    return state;
+  }
+};
+
+
+class SpiResource : public Resource {
+ public:
+  TAG(SpiResource);
+  SpiResource(ResourceGroup* group);
+  ~SpiResource();
+
+  int fd() { return fd_; }
+  void set_fd(int fd) {
+    ASSERT(fd_ == -1);
+    fd_ = fd;
+  }
+
+  int error() const { return error_; }
+  void set_error(int error) { error_ = error; }
+
+  uint8* buffer() { return buffer_; }
+
+  Object* transfer_start(Blob data, int from, int length,
+                         bool is_read,
+                         int delay_usecs,
+                         bool cs_change,
+                         Process* process);
+
+  Object* transfer_finish(bool was_read, Process* process);
+
+ private:
+  int fd_ = -1;
+  int error_ = 0;
+  AsyncEventThread* thread_ = null;
```

Comment:
Not sure if this would be the place to explain why we need a thread per resource.

Reply:
done.